File: service/src/main/java/mu/server/service/dto/Result.java
package mu.server.service.dto;

import lombok.Builder;
import org.jspecify.annotations.Nullable;

@Builder
public record Result<T>(@Nullable T value, @Nullable String error, boolean success) {

    public static <T> Result<T> ok(T value) {
        return new Result<>(value, null, true);
    }

    public static <T> Result<T> failure(String error) {
        return new Result<>(null, error, false);
    }
}

File: service/src/main/java/mu/server/service/dto/Result.kt
package mu.server.service.dto

data class Result<T>(val result: T? = null, val error: String? = null, val success: Boolean) {
    companion object {
        fun <T> ok(value: T): Result<T> = Result(value, null, true)

        fun <T> failure(error: String): Result<T> = Result(null, error, false)
    }
}
